### PR TITLE
Handle IntersectionType, UnionType, TypeAliasDeclaration, FirstType and stub ImportEqualsDeclaration

### PR DESCRIPTION
**Pull request by epabst**
This enables translating react.d.ts to Kotlin (after disabling exportedByAssignment check)!

### CODE REVIEW

File: src/ast/typescript/typeScriptAstUtils.kt [hunk 1]
```diff
@@ -222,6 +224,13 @@ fun TS.UnionTypeNode.toKotlinTypeName(typeMapper: ObjectTypeToKotlinTypeMapper):
     return DYNAMIC + commentWithExpectedType
 }
 
+fun TS.IntersectionTypeNode.toKotlinTypeName(typeMapper: ObjectTypeToKotlinTypeMapper): String {
+    val kotlinTypeNames = types.arr.map { it.toKotlinTypeName(typeMapper) }
+    val commentWithExpectedType = kotlinTypeNames.joinToString(" & ", prefix = " /* ", postfix = " */")
+    // just take the first one for now since Kotlin doesn't support intersection types.
+    return kotlinTypeNames[0] + commentWithExpectedType
```
Comment (bashor) [inline src/ast/typescript/typeScriptAstUtils.kt]: I think `dynamic` is better in this case

File: src/ast/typescript/typeScriptAstUtils.kt [hunk 2]
```diff
@@ -164,6 +164,8 @@ private fun TS.TypeNode.toKotlinTypeNameIfStandardType(typeMapper: ObjectTypeToK
 
         TS.SyntaxKind.ThisType -> (this.cast<TS.ThisTypeNode>()).toKotlinTypeName(typeMapper)
 
+        TS.SyntaxKind.FirstTypeNode -> (this.cast<TS.FirstTypeNode>()).toKotlinTypeName(typeMapper)
```
Comment (bashor) [inline src/ast/typescript/typeScriptAstUtils.kt]: How about tests? Which cases are covered by this?

File: src/TypeScriptToKotlinWalker.kt [hunk 3]
```diff
@@ -273,8 +273,10 @@ class TypeScriptToKotlinWalker(
                                 unsupportedNode(node)
                         }
 
-        exportedByAssignment[exportName] =
-                Annotation(MODULE, if (moduleName == null) listOf() else listOf(Argument(value = "\"$moduleName\"")))
+        // Kotlin doesn't need to export and import since it uses packages and visibility modifiers.
```
Comment (bashor) [inline src/TypeScriptToKotlinWalker.kt]: But how about tests? Sorry, but we can't disable it in the master. We should somehow map this TS construction to Kotlin.

File: src/ast/typescript/typeScriptAstUtils.kt (hunk 1 above)
Comment (epabst) [inline src/ast/typescript/typeScriptAstUtils.kt]: What benefits do you see for dynamic?  From what I can see, dynamic would provide zero type-safety, while checking the first one would provide at least partial type-safety.

File: src/TypeScriptToKotlinWalker.kt (hunk 3 above)
Comment (epabst) [inline src/TypeScriptToKotlinWalker.kt]: I'll undo this.

**Comment by epabst** [comment]
@bashor updated.

File: src/ast/typescript/typeScriptAstUtils.kt (hunk 2 above)
Comment (epabst) [inline src/ast/typescript/typeScriptAstUtils.kt]: Added.

File: src/ast/typescript/typeScriptAstUtils.kt (hunk 1 above)
Comment (bashor) [inline src/ast/typescript/typeScriptAstUtils.kt]: But to provide arguments with other types we should cast them.
Why first type is better?

**Comment by bashor** [comment]
Well, I've merged it. You can fix rest of notes in separate PR, if you want.
Anyway, probably we should/can create issues for them.

File: src/ast/typescript/typeScriptAstUtils.kt (hunk 1 above)
Comment (bashor) [inline src/ast/typescript/typeScriptAstUtils.kt]: Sorry, you right. In some reason I thought about union types...

Comment (bashor) [inline src/ast/typescript/typeScriptAstUtils.kt]: right now for functions we can emulate intersection types with [upper-bounds](https://kotlinlang.org/docs/reference/generics.html#upper-bounds), like:
```kotlin
fun <T> foo(): T where T : Number, T : Boo {

}
```

**Comment by epabst** [comment]
I'm wondering if this looks better.  What do you think?

fun <T : Number> foo(): T where T : Boo {

}

**Comment by bashor** [comment]
@epabst, yes, that's nice, I think I already suggested it somewhere w/o sample.

**Pull request merged**